File: 1636. Sort Array by Increasing Frequency.py
# Title:  1636. Sort Array by Increasing Frequency
# Idea:   O(n)/O(n+C)
# Complexity: O(n)/O(n+C)
# Language:Python
# Author:  Kaili Yang

# Solution1
# Python2
class Solution(object):
    def frequencySort(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        # 将列表转换成集合去重
        nums_set = set(nums)

        # 定义空字典和集合
        my_dict = {}
        my_list = []
        for item in nums_set:
            # 统计列表中每个元素出现的次数
            c = nums.count(item)

            # 将列表中的元素作为key,出现次数作为value,存到字典中
            my_dict[item] = c
        # 将字典按照value的大小升序排序,如果有多个值相同,按照key本身降序排序
        my_dict = sorted(my_dict.items(), key=lambda x: (x[1], x[0] * (-1)))
        for i in my_dict:
            for j in range(i[1]):
                # 将nums按照每个值的频率升序排序,如果有多个值的频率相同,按照数值本身将它们降序排序
                my_list.append(i[0])
        return my_list


# Solution2
# Python3
class Solution:
    def frequencySort(self, nums: List[int]) -> List[int]:
        return sorted(nums, key=lambda x:(mp[x], -x)) if (mp := Counter(nums)) else nums

File: 1636. Sort Array by Increasing Frequency.java
// Title:  1636. Sort Array by Increasing Frequency
// Idea:   O(n)/O(n+C)
// Complexity: O(n)/O(n+C)
// Language: Java
// Author:  Kaili Yang

// 按照题意，统计词频，再按词频排序即可。
// 注意在Java中是不支持对基本类型的数组自定义排序的，所以，这里我们需要要用Integer[] 数组。

class Solution {
    public int[] frequencySort(int[] nums) {
        Integer[] arr = new Integer[nums.length];
        int[] freqs = new int[201];
        for (int i = 0; i < nums.length; i++) {
            // 统计频次
            freqs[nums[i] + 100]++;
            //将int数组转换成Integer数组
            arr[i] = nums[i];
        }
        // Java中不支持对基本类型的数组自定义排序
        Arrays.sort(arr, (a, b) -> {
            // 按照频次排序
            return freqs[a + 100] == freqs[b + 100] ? b - a : freqs[a + 100] - freqs[b + 100];
            
            // 如果两个数的频次相同，那么这两个数就按照从大到小排序（后者减去前者）
            // 频次不相同，则按照频次排序
        });
        // 转换为int数组
        for (int i = 0; i < nums.length; i++) {
            nums[i] = arr[i];
        }
        return nums;
    }
}

// Soluion2
class Solution {
    public int[] frequencySort(int[] nums) {
        int n = nums.length;
        Map<Integer,Integer> map = new HashMap<>();
        for(int i = 0;i<n;i++){
            map.put(nums[i],map.getOrDefault(nums[i],0)+1);
        }
        List<int[]> list = new ArrayList<>();
        for(int key:map.keySet()){
            list.add(new int[]{key,map.get(key)});
        }
        Collections.sort(list,(a,b)->{
            if(a[1] != b[1]) return a[1] - b[1];
            return b[0] - a[0];
        });
        int[] ans = new int[n];
        int idx = 0;
        for(int[] cur:list){
            int m = cur[1];
            while(m-->0) ans[idx++] = cur[0];
        }
        return ans;
    }
}